scheduler/src/cook/kubernetes/api.clj

```diff
@@ -0,0 +1,316 @@
+                  {:state :pod/succeeded
+                   :exit exit-code
+                   :reason (-> state .getTerminated .getReason)}
+                  {:state :pod/failed
+                   :exit exit-code
+                   :reason (-> state .getTerminated .getReason)}))
+              :default
+              {:state :pod/unknown
+               :reason "Unknown"}))
+
+          {:state :pod/waiting
+           :reason "Pending"})))))
+
+(defn kill-task
+  "Kill this kubernetes pod"
+  [^ApiClient api-client ^V1Pod pod]
+  (let [api (CoreV1Api. api-client)
+        ^V1DeleteOptions deleteOptions (-> (V1DeleteOptionsBuilder.) (.withPropagationPolicy "Background") .build)]
+    (try
+      (.deleteNamespacedPod
+        api
+        (-> pod .getMetadata .getName)
+        (-> pod .getMetadata .getNamespace)
+        deleteOptions
+        nil
+        nil
+        nil
+        nil
+        nil)
+      (catch JsonSyntaxException e
+        (log/info "Caught the https://github.com/kubernetes-client/java/issues/252 exception.")
+        ; Silently gobble this exception.
+        ;
+        ; The java API can throw a a JsonSyntaxException parsing the kubernetes reply!
+        ; https://github.com/kubernetes-client/java/issues/252
+        ; https://github.com/kubernetes-client/java/issues/86
+        ;
+        ; They actually advise catching the exception and moving on!
+        ;
+        ))))
+
+
+(defn remove-finalization-if-set-and-delete
+  "Remove finalization for a pod if its there. No-op if its not there.
+  We always delete unconditionally, so finalization doesn't matter."
+  [^ApiClient api-client ^V1Pod pod]
+  ; TODO: This likes to noisily throw NotFound multiple times as we delete away from kubernetes.
+  ; I suspect our predicate of existing-states-equivalent needs tweaking.
+  (kill-task api-client pod))
+
+(defn launch-task
+  "Given a pod-name use lookup the associated task, extract the parts needed to syntehsize the kubenretes object and go"
+  [api-client {:keys [launch-pod] :as expected-state-dict}]
+  ;; TODO: make namespace configurable
+  (let [namespace "cook"]
+    ;; TODO: IF there's an error, log it and move on. We'll try again later.
+    (if launch-pod
+      (let [api (CoreV1Api. api-client)]
+        (log/info "Launching pod" api launch-pod)
+        (try
+          (-> api
+              (.createNamespacedPod namespace launch-pod nil nil nil))
+          (catch ApiException e
+            (log/error e "Error submitting pod:" (.getResponseBody e)))))
+      ; Because of the complicated nature of task-metadata-seq --- we can't easily run the V1Pod creation code for a
+      ; failed-to-start pod on a server restart. Thus, if we create a task, store into datomic, but then the cook scheduler
+      ; fails --- before kubernetes creates a pod (either the message isn't sent, or there's a kubernetes problem), we will
+      ; the inability to create a new V1Pod and we can't retry this at the kubernetes level.
```

Comment:
Grammar doesn't make sense "... we will the inability ..."

Reply:
Reised